packages/test-base/src/commonTest/kotlin/io/realm/kotlin/test/common/RealmObjectTests.kt

```diff
@@ -113,6 +141,173 @@ class RealmObjectTests : RealmStateTest {
+        assertTrue(regex.matches(managedObj.toString()), managedObj.toString())
+    }
+
+    @Test
+    fun toString_managed_cyclicData() {
+        val p1 = SampleWithPrimaryKey()
+        p1.stringField = "Parent"
+        p1.nullableObject = p1
+        val managedObj = realm.writeBlocking { copyToRealm(p1) }
+        val regex = Regex("io.realm.kotlin.entities.SampleWithPrimaryKey\\{state=VALID, schemaName=SampleWithPrimaryKey, objKey=[0-9]*, version=[0-9]*, realm=${realm.configuration.name}\\}")
+        assertTrue(regex.matches(managedObj.toString()), managedObj.toString())
+    }
+
+    @Test
+    fun toString_managed_invalid() {
+        realm.writeBlocking {
+            val managedObject = copyToRealm(Parent())
+            delete(managedObject)
+            val regex = Regex("io.realm.kotlin.entities.link.Parent\\{state=INVALID, schemaName=Parent, realm=${realm.configuration.name}, hashCode=[-0-9]*\\}")
+            assertTrue(regex.matches(managedObject.toString()), managedObject.toString())
+            cancelWrite()
+        }
+    }
+
+    @Test
+    fun toString_managed_closedRealm() {
+        val managedObject = realm.writeBlocking {
+            copyToRealm(Parent())
+        }
+        realm.close()
+        val regex = Regex("io.realm.kotlin.entities.link.Parent\\{state=CLOSED, schemaName=Parent, realm=${realm.configuration.name}, hashCode=[-0-9]*\\}")
+        assertTrue(regex.matches(managedObject.toString()), managedObject.toString())
+    }
+
+    @Test
+    fun toString_customMethod() {
+        assertEquals("customToString", CustomMethods().toString())
+        val managedObj = realm.writeBlocking { copyToRealm(CustomMethods()) }
+        assertEquals("customToString", managedObj.toString())
+    }
+
+    @Test
+    fun toString_unmanaged() {
+        val unmanagedObject = Parent()
+        val regex = Regex("io.realm.kotlin.entities.link.Parent\\{state=UNMANAGED, schemaName=Parent, hashCode=[-0-9]*\\}")
+        assertTrue(regex.matches(unmanagedObject.toString()), unmanagedObject.toString())
+    }
+
+    @Test
+    fun equals_hashCode_managed() {
+        realm.writeBlocking {
+            val p1 = copyToRealm(Parent().apply { this.name = "Jane" })
+            val p2 = copyToRealm(Parent())
+            val p3 = query(Parent::class, "name = 'Jane'").first().find()!!
+            assertEquals(p1, p1)
+            assertEquals(p1, p3)
+            assertEquals(p1.hashCode(), p1.hashCode())
+            assertEquals(p1.hashCode(), p3.hashCode())
+
+            // Not restrictions are given on hashCode if two objects are not equal
+            assertNotEquals(p2, p1)
+            assertNotEquals(p2, p3)
+        }
+    }
+
+    @Test
+    fun equals_hashCode_unmanaged() {
+        val p1 = Parent()
+        val p2 = Parent()
+        assertEquals(p1, p1)
+        assertEquals(p1.hashCode(), p1.hashCode())
+        assertNotEquals(p1, p2)
+    }
+
+    @Test
+    fun equals_hashCode_mixed() {
```

Comment:
Nothing related to `hashCode` is asserted. If we want to keep the name, then maybe add the your comment that we cannot guarantee anything for hashes on unequal objects to make it clear that we actually thought of it.